### PR TITLE
fix: save logo caption and description when uploading app

### PR DESCRIPTION
**Pull request by Birkbjo**
Currently the information about the image uploaded together with the app was ignored. This adds support to add caption and description when creating a new app.

### CODE REVIEW

File: src/routes/v1/apps/handlers/createApp.js [hunk 1]
```diff
@@ -223,14 +223,20 @@ module.exports = {
...
+                    ;({ caption, description } = imageInfo)
+                }
                 const { id, uuid } = await addAppVersionMedia(
                     {
                         userId: requestUserId,
                         appVersionId: appVersion.id,
                         imageType: ImageType.Logo,
                         fileName: imageFileMetadata.filename,
                         mime: imageFileMetadata.headers['content-type'],
+                        caption: caption,
```
Comment (Birkbjo) [inline src/routes/v1/apps/handlers/createApp.js]: Note that these can be undefined. Shorthand would work, but would give `undefined: undefined`, which is not prefereable.

File: src/routes/v1/apps/handlers/createApp.js [hunk 2]
```diff
@@ -223,14 +223,20 @@ module.exports = {
...
+                        description: description,
```
Comment (Birkbjo) [inline src/routes/v1/apps/handlers/createApp.js]: Note that these can be undefined. Shorthand would work, but would give undefined: undefined, which is not prefereable.

File: src/data/addAppVersionMedia.js [hunk 3]
```diff
@@ -96,6 +98,8 @@ const addAppVersionMedia = async (params, knex, transaction) => {
             created_by_user_id: userId,
             media_type_id: mediaTypeId,
             app_version_id: appVersionId,
+            caption: params.caption,
+            description: params.description,
```
[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data/addAppVersionMedia.js]: They are nested in `payload.images`-array, so that will fail if no image is passed, right?

**Comment by erikarenhill** [comment]
I can't actually get this to work when testing, the frontend doesn't seem to add the caption or description to the json payload within app.images that the backend reads, only an empty array.

It seems that this code always overwrites the images array as imageFile.length is never > 0 (it can be either empty array or a single image object):

https://github.com/dhis2/app-store/blob/master/client/src/components/form/UploadAppFormStepper.jsx#L303

**Comment by erikarenhill** [comment]
yes, tested locally to merge with latest master and then it works.

> I can't actually get this to work when testing, the frontend doesn't seem to add the caption or description to the json payload within app.images that the backend reads, only an empty array.
> 
> It seems that this code always overwrites the images array as imageFile.length is never > 0 (it can be either empty array or a single image object):
> 
> However in the master branch this seems to be fixed as it only checks for imageFile not being falsy
> instead of imageFile.length

yes, tested locally to merge with latest master and then it works.

**Comment by erikarenhill** [comment]
@Birkbjo would be nice with a unit test for this as well

File: test/data/index.js [hunk 4]
```diff
@@ -20,6 +21,34 @@ describe('@data::addAppVersionMedia', () => {
...
+            userId: 1, //travis user
+            imageType: ImageType.Screenshot,
+            fileName: 'screenshot.jpg',
+            mime: 'image/jpeg',
+            caption: 'Test caption',
+            description: 'a description',
+        }
+        const trx = await db.transaction()
+        const { id, uuid } = await addAppVersionMedia(appMedia, db, trx)
+        trx.commit()
```
[the remark this answers is not in the document]
Comment (erikarenhill) [reply test/data/index.js]: need to revisit this with regards to the automatic rollback mechanism

**Pull request merged**